core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -3497,6 +3498,13 @@ object ZIO extends ZIOCompanionPlatformSpecific {
       self.provideLayer[E1, R0, R0 with R1](ZLayer.identity[R0] ++ layer)
   }
 
+  @implicitNotFound(
```

Comment:
We discovered you can "Inline" this annotation to avoid the need for a new `CanFilter` class.

Reply:
Like that?
```
object Evidences {
  @implicitNotFound("provideCustomLayer expect all layers except those provided by ZIO (ZEnv) instead it got $R1. If you intend to provide some of layers only then use provideSomeLayer instead")
  type ProvideCustomLayer[-R1, +R] = ZEnv with R1 <:< R
}
```

Reply:
Even more! Inline where the function is:

```scala
@implicitNotFound("") implicit ev: ...
```

Try that!

Reply:
Note that this applies to 2.13+ only (which may be fine anyway)

Reply:
Yeah, I tried it and it doesn't work on default scala version. 
Is there any runtime overhead? If no - I'd go with a bulky solution.
If yes - we can put extensions in version-specific directories

Reply:
@jdegoes @neko-kai I made implementation version-specific. Have no idea what I'm doing but it compiles!